keps/0007-command-asserts.md

```diff
@@ -0,0 +1,103 @@
+  - [Implementation Details](#implementation-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+By asserting CLI commands, kuttl can support many additional test scenarios that examing the internal state of an operator.
+
+## Motivation
+
+Operators and custom resources can be used to deploy complex applications. For example, KUDO provides operators for Apache Cassandra and Apache Kafka. Deploying custom resources of these operators may not always result in changed Kubernetes resources but also in changes that aren't surfaced by Kubernetes resources but are internal to the operator. Asserting these changes only works through accessing service APIs or running commands and parsing their output. kuttl should provide tools to assert such output.
+
+### Goals
+
+- Assert that commands examining the internal state of an operator succeed or fail
+- Keep eventual consistency when asserting commands
+
+### Non-Goals
+
+- Introduce "flows" to determine when to run commands. I.e. to run commands once before or after asserting resources
+
+## Proposal
+
+Introduce commands as part of `TestAssert`. These commands are run at a regular interval like resource assertions until they either succeed or a timeout of the assert is reached. If any of the commands fail, the assert is considered to have failed. A command's return value determines if a command failed or not.
+
+As a start, complex commands that parse output can be described using shell environments. At a later stage we can implement additional tools to simplify common tasks. E.g. to search with a regular expression in command output.
+
+### User Stories
+
+#### User Story 1
+
+The JVM options of a KUDO operator can be tuned using parameters. On deployment, the JVM options are logged in the pods running the application provided by the operator. To assert that a parameter change resulted in changed JVM options we need to run `kubectl logs pod-name` and search the logs for a specific string once the operator has been deployed.
+
+#### User Story 2
+
+An operator can be configured to encrypt it's API. To assert that the API endpoints are indeed encrypted we can run a specific `curl` command and check that it returns successfully once the operator has been deployed.
+
+### Implementation Details
+
+Command assertions use most fields of the existing commands for `TestStep` and follow their behavior. The `ignoreFailure`, `background` and `timeout` fields are removed as they are not needed here.
+
+An example that parses the logs of a pod:
+
+```yaml
+apiVersion: kuttl.dev/v1beta1
+type: TestAssert
+commands:
+  - command: ./parse-logs.sh name-of-pod-0 string-to-search-for
+    namespaced: true
+```
+
+with `parse-logs.sh`:
+
+```bash
+#!/bin/bash
+
+POD=$1
+SEARCH=$2
+# $3 is "--namespace"
+NAMESPACE=$4
+
+kubectl logs -n ${NAMESPACE} ${POD} | grep ${SEARCH}
+```
+
```

Comment:
I would be inclined to cycle through all cmd asserts first until they are successful... then run all declarative asserts repeatedly until complete.  (they all have to be true to succeed anyway)

Also... if someone wants to have a non-connected assert... they can always have 00-assert with declarative and 01-assert with commands.

Reply:
My idea is to cycle through the commands it the same loop that cycles through the objects. I'll detail this in the KEP. This would keep the implementation simple. Asserting commands before or after asserting Kubernetes objects could be added later.